trio/_subprocess.py

```diff
@@ -389,24 +391,47 @@ def kill(self):
        process group.
 
     """
-    default_redirect = None if passthrough else subprocess.PIPE
-    options.setdefault("stdin", default_redirect)
-    options.setdefault("stdout", default_redirect)
-    options.setdefault("stderr", default_redirect)
 
-    if input is not None and options["stdin"] != subprocess.PIPE:
+    if isinstance(stdin, str):
+        raise UnicodeError("process stdin must be bytes, not str")
+    if stdin == subprocess.PIPE:
         raise ValueError(
-            "can't provide input to a process whose stdin is redirected"
+            "stdin=subprocess.PIPE doesn't make sense since the pipe "
+            "is internal to run_process(); pass the actual data you "
+            "want to send over that pipe instead"
         )
-
-    if options["stdin"] == subprocess.PIPE and not input:
-        options["stdin"] = subprocess.DEVNULL
+    if isinstance(stdin, bytes):
+        input = stdin
+        if input:
+            options["stdin"] = subprocess.PIPE
+        else:
+            options["stdin"] = subprocess.DEVNULL
```

Comment:
I can, but if you want DEVNULL instead of an empty pipe, then use DEVNULL. Explicit is better than premature micro-optimization, or something like that.